iothub/device/src/Pipeline/TransportDelegatingHandler.cs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Transport
+{
+    /// <summary>
+    /// Transport delegating handler.
+    /// This handler is responsible for initializing the transport handler on initial connection and subsequent reconnects.
+    /// </summary>
+    internal class TransportDelegatingHandler : DefaultDelegatingHandler
+    {
+        private SemaphoreSlim _handlerLock = new(1, 1);
+
+        public TransportDelegatingHandler(PipelineContext context, IDelegatingHandler innerHandler)
+            : base(context, innerHandler)
+        {
+        }
+
+        public override async Task OpenAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                if (Logging.IsEnabled)
+                    Logging.Enter(this, cancellationToken, $"{nameof(TransportDelegatingHandler)}.{nameof(OpenAsync)}");
+
+                cancellationToken.ThrowIfCancellationRequested();
+                await _handlerLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    CreateNewTransportIfNotReady();
+                    await base.OpenAsync(cancellationToken).ConfigureAwait(false);
+
+                    // since Dispose is not synced with _handlerLock, double check if disposed.
+                    if (_disposed)
+                    {
+                        NextHandler?.Dispose();
+                        ThrowIfDisposed();
+                    }
+                }
+                finally
+                {
+                    _handlerLock.Release();
+                }
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, cancellationToken, $"{nameof(TransportDelegatingHandler)}.{nameof(OpenAsync)}");
+            }
+        }
+
+        private void CreateNewTransportIfNotReady()
+        {
+            if (NextHandler == null || !NextHandler.IsUsable)
+            {
+                IDelegatingHandler innerHandler = NextHandler;
+
+                // Ask the ContinuationFactory to attach the proper handler given the Context's ITransportSettings.
+                NextHandler = ContinuationFactory(Context, null);
+
+                innerHandler?.Dispose();
+            }
+        }
+
+        public override async Task WaitForTransportClosedAsync()
```

Comment:
This logic is required to recover the MQTT transport handler on reconnections. It doesn't look like AMQP relies on this logic though. Since we can get rid of this implementation once we move to MQTTNet, I'm not spending time optimizing this code.